### PR TITLE
refactor: Support for component instance styling

### PR DESCRIPTION
**Pull request by MarcinVaadin**
## Description
Added support for applying style rules to given component instance.

### Update to CssRule structure
Selector has been replaced by tagName and partName.

### Applying styles to given instance
Request must contain nodeId, example:

themeEditorRules:
```
{
 uiId: 0,
 nodeId: 0
 rules: [
   {
     tagName: 'vaadin-button',
     partName: 'label',
     properties: {
        color: 'red'
     }
  }]
}
```

themeEditorLoadRules:
```
{
 uiId: 0,
 nodeId: 0
 selectorFilter: 'vaadin-button' // optional when nodeId is specified
}
```

After applying unique classname application must be rebuild and redeployed (or hotdeployed).

### CODE REVIEW

File: vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/JavaSourceModifier.java [hunk 1]
```diff
@@ -102,6 +105,60 @@ public void setClassNames(Integer uiId, Integer nodeId,
...
+        try {
+            ClassNameHolder holder = new ClassNameHolder();
+            VaadinSession session = getSession();
+            getSession().access(() -> {
+                Component component = getComponent(session, uiId, nodeId);
+                ComponentTracker.Location location = getComponentLocation(
+                        component);
+
+                File sourceFolder = getSourceFolder(location);
+                File sourceFile = new File(sourceFolder, location.filename());
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/JavaSourceModifier.java:130]: @sonatype-lift ignore

File: vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/JavaSourceModifier.java [hunk 2]
```diff
@@ -239,6 +303,29 @@ protected VaadinSession getSession() {
...
+        return n.getStatements().stream()
+                .filter(ExpressionStmt.class::isInstance)
+                .map(ExpressionStmt.class::cast)
+                .filter(e -> e.getExpression().isMethodCallExpr()) // filter
+                                                                   // method
+                                                                   // calls
+                .filter(e -> Objects.equals(variableName, getVariableName(e))) // filter
+                                                                               // variable
+                                                                               // scope
+                .filter(e -> Objects.equals("addClassName",
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/JavaSourceModifier.java:317]: @sonatype-lift ignore

File: vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/ThemeModifier.java [hunk 3]
```diff
@@ -111,16 +106,7 @@ public List<CssRule> getCssRules(Predicate<String> selectorFilter) {
                 .filter(rule -> rule.getSelectorCount() > 0)
                 .filter(rule -> selectorFilter
                         .test(rule.getSelectorAtIndex(0).getAsCSSString()))
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/ThemeModifier.java:108]: @sonatype-lift ignore

File: vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/ThemeModifier.java [hunk 4]
```diff
@@ -323,4 +316,39 @@ protected String getThemeName(File themes) {
...
+        rule.getAllDeclarations()
+                .forEach(cssDeclaration -> properties.put(
+                        cssDeclaration.getProperty(),
+                        cssDeclaration.getExpressionAsCSSString()));
+        return new CssRule(tagName, partName, className, properties);
+    }
+
+    protected String getTagName(CSSSelector selector) {
+        // assume tag name is always first
+        return selector.getMemberAtIndex(0).getAsCSSString();
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/ThemeModifier.java:334]: @sonatype-lift ignore

**Comment by MarcinVaadin** [comment]
Moved component accessible flag from ComponentMetadata request to LoadRulesResponse. If component is not accessible unique classname would not be checked and applied.

File: vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/ThemeModifier.java [hunk 5]
```diff
@@ -71,11 +68,9 @@ public void setThemeProperties(List<CssRule> rules) {
                     .entrySet()) {
                 if (property.getValue() != null
                         && !property.getValue().isBlank()) {
-                    setCssProperty(styleSheet, rule.getSelector(),
-                            property.getKey(), property.getValue());
+                    setCssProperty(styleSheet, rule);
```
[the remark this answers is not in the document]
Comment (MarcinVaadin) [reply vaadin-dev-server/src/main/java/com/vaadin/base/devserver/themeeditor/ThemeModifier.java]: Fixed

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**